pkg/distributor/distributor.go

```diff
@@ -908,18 +893,41 @@ func (d *Distributor) PushWithCleanup(ctx context.Context, req *mimirpb.WriteReq
 		return d.send(localCtx, ingester, timeseries, metadata, req.Source)
 	}, func() { cleanup(); cancel() })
 
-	if forwardingErrCh != nil {
-		// Blocks until the forwarding requests have completed and the final status has been pushed through this chan.
-		forwardingErr := <-forwardingErrCh
-		err = httpgrpcutil.PrioritizeRecoverableErr(err, forwardingErr, firstPartialErr)
-	}
-
 	if err != nil {
 		return nil, err
 	}
 	return &mimirpb.WriteResponse{}, firstPartialErr
 }
 
+// forwardingReq returns a forwarding request if one is necessary, given the user ID.
+// if no forwarding request is necessary it returns nil.
+func (d *Distributor) forwardSamples(ctx context.Context, req *mimirpb.WriteRequest, userID string, cleanup func()) (*mimirpb.WriteResponse, error) {
+	forwardingRules := d.limits.ForwardingRules(userID)
+	if len(forwardingRules) == 0 {
+		return d.PushWithCleanup(ctx, req, cleanup)
+	}
+
+	var forwardingErrCh <-chan error
+	var notIngestedCounts forwarding.TimeseriesCounts
+
+	// Reassign req.Timeseries because the forwarder creates a new slice which has been filtered down.
+	// The cleanup func will cleanup the new slice, it's the forwarders responsibility to return the old one to the pool.
+	notIngestedCounts, req.Timeseries, forwardingErrCh = d.forwarder.Forward(ctx, forwardingRules, req.Timeseries)
+
+	// Some samples have been forwarded and won't be ingested but we still need to account for them in some of the metrics.
+	d.incomingSamples.WithLabelValues(userID).Add(float64(notIngestedCounts.SampleCount))
+	d.incomingExemplars.WithLabelValues(userID).Add(float64(notIngestedCounts.ExemplarCount))
+
+	resp, err := d.PushWithCleanup(ctx, req, cleanup)
+
+	errs := []error{err}
+	for err := range forwardingErrCh {
+		errs = append(errs, err)
+	}
```

Comment:
I don't know where to write this comment, I think this is the best place: if all workers are busy or in a deadlock or whatever, this will block forever even if the context was canceled.

Reply:
good catch! this is fixed now:

https://github.com/grafana/mimir/blob/1373bff99be70bbb9433ce26b9b53aac6eb28b21/pkg/distributor/distributor.go#L630-L644